third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerWarpGroup.cpp

```diff
@@ -102,31 +100,78 @@ class LowerWarpGroup : public OpRewritePattern<WarpGroupOp> {
       mlir::getUsedValuesDefinedAbove(*partition, captures);
 
     SmallVector<Value> inputs;
-    SmallVector<Value> constants;
-    for (Value capture : captures) {
-      if (!isa<BlockArgument>(capture) &&
-          isa<arith::ConstantOp>(capture.getDefiningOp())) {
-        constants.push_back(capture);
+    SmallVector<IRMapping> mappings(partitions.size());
+    SmallVector<OpBuilder> builders;
+    for (auto region : partitions) {
+      builders.push_back(OpBuilder::atBlockBegin(&region->front()));
+    }
+
+    SetVector<Operation *> opsToClone;
+    std::queue<Value> que;
+    for (auto capture : captures) {
+      que.push(capture);
+    }
+
+    while (!que.empty()) {
+      Value capture = que.front();
+      Operation *defOp = capture.getDefiningOp();
```

Comment:
can you also move over the comments from PartitionLoops.cpp describing the restriction and rationale here?